### ISSUE TITLE
Getting an error after adding "renderNodes"

### ISSUE THREAD

**Issue by kabugh** [paramander/contentful-rich-text-vue-renderer#9]
Hi, I am using vue + typescript configuration.  When I try to add renderNodes object to `<RichTextRenderer />`, I get an error:

![Screenshot 2020-02-28 at 10 56 17](https://user-images.githubusercontent.com/51855058/75538498-f71fe900-5a18-11ea-8984-dd6fb6f91567.png)

Here's an object I get from my contentful space which gets passed to `:document`: 

![Screenshot 2020-02-28 at 10 49 14](https://user-images.githubusercontent.com/51855058/75538638-3817fd80-5a19-11ea-8137-0fb51aee2ff7.png)

The custom renderer I got from [rich-text-html](https://github.com/contentful/rich-text/issues/96 ) should replace `\n` with line break. (doesn't work here)

If I use official rich-text-html and pass renderNodes, it does work somehow, so I guess it must be a vue-plugin-related problem.

My code:
```
<template>
    <div>
        <RichTextRenderer
          :document="myData.fields.description"
          :nodeRenderers="renderNodes"
        />
    </div>
</template>
<script lang="ts">
    import { Component, Vue } from "vue-property-decorator";
    import { BLOCKS } from "@contentful/rich-text-types";
    import RichTextRenderer from "contentful-rich-text-vue-renderer";

    const renderNodes = {
      [BLOCKS.PARAGRAPH]: (node, next) =>
        `<p>${next(node.content).replace(/\n/g, "<br/>")}</p>`
    };

    @Component({
      components: {
        RichTextRenderer
      }
    })
    export default class MyComponent extends Vue {}
</script>
```

Expected output: The `renderNodes` get passed and the renderer is used correctly.

What could go wrong?

**Comment by tolgap** [comment]
@kabugh You create a `const renderNodes` that is only available within the scope of your script tag. Not in the template.

Templates can only reference variables that are bound to your component. So you can move the `renderNodes` variable to `MyComponent.data` or to `MyComponent.computed`.

**Comment by kabugh** [comment]
@tolgap Thanks, that was it.. but after doing so Typescript complains that `next` is not a function, even though I set its type explicitly: `[BLOCKS.PARAGRAPH]: (node: any, next: Function)`.

**Comment by tolgap** [comment]
@kabugh This is the current definition of the default paragraph renderer. I'm not quite sure where you got your notation from, but it receives 4 params.

* `h` is the Vue component renderer
* `key` is required by Vue's own renderer. Make sure to add the key just like below. 
* `next` is the function that will further render any nodes/marks inside the `p` tag. Make sure the wrap the result of your function into next.

```js
const defaultNodeRenderers = {
    [BLOCKS.PARAGRAPH]: (node, key, h, next) => (
        h('p', { key }, next(node.content, key, h, next))
    ),
}
```

So with the explanation above, your paragraph renderer should look like

```js
{
  [BLOCKS.PARAGRAPH]: (node, key, h, next) => {
    const newlineToBr = node.content.replace(/\n/g, "<br/>")
    return h('p', { key }, next(newlineToBr, key, h, next));
  }
};
```

You might have to replace the `\n` with a Vue rendered `<br>` though. So you will have to pass `<br/>` through the `h` function as well.

**Comment by kabugh** [comment]
I am not really familiar with "behind the scene" of vue. After trying the new solution, I get an error: `[Vue warn]: Error in render: "TypeError: node.content.replace is not a function"`. Also tried to call `replace()` on next as I got it from `rich-text-html`, but it also doesn't work as intended.

```javascript
[BLOCKS.PARAGRAPH]: (node, key, h, next) => {
      const newlineToBr = node.content;
      return h(
        "p",
        { key },
        next(newlineToBr, key, h, next).replace(/\n/g, "<br/>")
      );
    }
```

**Comment by tolgap** [comment]
@kabugh You have a very specific issue with the **_data_** you're feeding into the documentRenderer. Can you please paste the specific paragraph JSON that is causing this issue?

**Comment by kabugh** [comment]
@tolgap Oh, so I passed the whole object as shown in my first post, now I tried passing only a single paragraph and indeed it gets rendered, however the line breaks `<br />` don't work. That might be the issue?

Here's my data which I am passing to `<RichTextRenderer />`

![Screenshot 2020-02-28 at 13 18 28](https://user-images.githubusercontent.com/51855058/75548146-d792bb80-5a2c-11ea-86d3-9087321a3054.png)

**Comment by tolgap** [comment]
@kabugh I would love to help you, but I just don't have the time to explain how within Github Issues.

1. I need to explain how Vue works with Virtual Dom (the `h` function specifically)
2. How contentful's data is controlled
3. The fact that you are assuming that `node.content` is a String, when it's an Array (hence the replace function error)
4. The `\n` is rendered as a text node, not as a paragraph

This issue requires so much specific work to fix when working with Vue, that I hope you can figure it out and paste it here. I would love to add a `newline to <br>` renderer as an example in the README. More people might run into this issue.

**Comment by damienroche** [comment]
@kabugh have you solved this problem ?

**Comment by kabugh** [comment]
@damienroche  nah, not exactly, I sticked to the solution that I override contentful horizontal line and replace it with `<br>` since I don't need it anyways.

`[BLOCKS.HR]: (node, next) => '<br>'`

**Comment by tolgap** [comment]
@damienroche @kabugh 

I just dove into this. I managed to solve it by introducing a custom node renderer for `<br>`:

```javascript
      const nodeRenderers = {
        break: (_node, key, h) => h('br', key, {}), // <-- a custom break node renderer

        // we override the paragraph renderer. We split every text node inside the paragraph on `\n`.
        // And then we can "intersperse" these text nodes with our new custom `<br>` node.
        [BLOCKS.PARAGRAPH]: (node, key, h, next) => {
          const nodeContentWithNewlineBr = node.content.map(childNode => {
            if (childNode.nodeType === 'text') {
              const splittedValue = childNode.value.split("\\n");
              return splittedValue.reduce((aggregate, v, i) => (
                [
                  ...aggregate,
                  { ...childNode, value: v },
                  { nodeType: 'break', key: `${key}-br-${i}` }
                ]
              ), []).slice(0, -1);
            }

            return childNode;
          });

          node.content = [].concat.apply([], nodeContentWithNewlineBr);

          return h('p', { key }, next(node.content, key, h, next));
        }
      };
```

Here is the full test for this: https://github.com/paramander/contentful-rich-text-vue-renderer/blob/master/src/index.test.js#L389

and the jest expectation: https://github.com/paramander/contentful-rich-text-vue-renderer/blob/master/src/index.test.js#L453

**Comment by mischaboldy** [comment]
I realise this is closed, but wanted to add the following (in case it might help someone)

I used the above code from @tolgap (thanks btw!) but got a vue warning:

`You may have an infinite update loop in a component render function.`

This is because when doing:

`node.content = [].concat.apply([], nodeContentWithNewlineBr);`

it updates state which triggers a rerender in vue creating an infinite loop. So I found an easy fix is to change the last lines from:

```
node.content = [].concat.apply([], nodeContentWithNewlineBr);

return h('p', { key }, next(node.content, key, h, next));
```

to:

`return h('p', { key }, next([].concat.apply([], nodeContentWithNewlineBr), key, h, next))`

**Comment by tolgap** [comment]
@mischaboldy This was fixed in #34. It's a small change. Your solution is basically the same 😁.
